src/charm.py

```diff
@@ -110,63 +102,191 @@ def _missing_mandatory_relations(self) -> Optional[str]:
+            # event code from running. Luckily, for this charm, none of the
+            # teardown code is necessary to perform if we're removing the
+            # charm.
+            event.add_status(BlockedStatus("Scaling is not implemented for this charm"))
+            return
+
         if missing_relation := self._missing_mandatory_relations():
-            self.unit.status = BlockedStatus(
-                f"Waiting for `{missing_relation}` relation to be created"
-            )
+            event.add_status(BlockedStatus(f"Waiting for {missing_relation} relation"))
             return
+
         if not self._container.can_connect():
-            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.add_status(WaitingStatus("Waiting for container to be ready"))
             return
+
         if not self._database_is_available():
-            self.unit.status = WaitingStatus("Waiting for `database` relation to be available")
+            event.add_status(WaitingStatus("Waiting for `database` relation to be available"))
             return
+
         if not self._nrf_is_available():
-            self.unit.status = WaitingStatus("Waiting for NRF relation to be available")
+            event.add_status(WaitingStatus("Waiting for NRF relation to be available"))
             return
+
         if not self._storage_is_attached():
-            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.add_status(WaitingStatus("Waiting for storage to be attached"))
+            return
+
+        if not _get_pod_ip():
+            event.add_status(WaitingStatus("Waiting for pod IP address to be available"))
+            return
+
+        if self._csr_is_stored() and not self._get_current_provider_certificate():
+            event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            return
+
+        if not self._smf_service_is_running():
+            event.add_status(WaitingStatus("Waiting for SMF service to start"))
             return
+
+        event.add_status(ActiveStatus())
+
+    def _smf_service_is_running(self) -> bool:
+        """Check if the SMF service is running.
+
+        Returns:
+            bool: Whether the SMF service is running.
+        """
+        if not self._container.can_connect():
+            return False
+        try:
+            service = self._container.get_service(self._service_name)
+        except ModelError:
+            return False
+        return service.is_running()
+
+    def ready_to_configure(self) -> bool:
+        """Returns whether the preconditions are met to proceed with the configuration.
+
+        Returns:
+            ready_to_configure: True if all conditions are met else False
+        """
+        if self._missing_mandatory_relations():
+            return False
+
+        if not self._container.can_connect():
+            return False
```

Comment:
I'm not sure about this check here. I think we'd be better off checking that directly before any action that requires the container to be accessible. Having this here kinda gives us the false sense of security.
Could be a topic for a different task tho. WDYT?

Reply:
I added this check as a first one before starting configuration (in the main callback function). Do you think it is enough ?

Reply:
Nope, I don't think so. My point is that we should be doing this check as close to the actual action as possible. We'll probably need to call it a lot in the code.
Right now, we do the check before doing other readiness checks. By the time we get to the actual configuration of the workload, the output of `can_connect` will be very much outdated.

Reply:
IMO, this could be another task. Maybe we could use a decorator which checks can_connect before every action.

Reply:
Nope, I'll do it.